psycopg/psycopg/_connection_base.py

```diff
@@ -313,6 +321,15 @@ def _try_cancel(cls, pgconn: "PGconn") -> None:
         else:
             c.cancel()
 
+    def _cancel_gen(self) -> PQGenConn[None]:
+        try:
+            cancel_conn = self.pgconn.cancel_conn()
+        except e.OperationalError as ex:  # if the connection is closed
+            logger.warning("couldn't create a cancel connection: %s", ex)
```

Comment:
Why to catch an exception, and assume that it was caused by the connection closed? We can just check if the connection is close if we want to verify that. It actually seems that we do it already, in virtue of the fact that all the calls to this method are guarded by `_should_cancel()`.

I don't see a strong reason to catch exceptions here. I don't think you want to catch *any* exception, because you don't catch them in the two statements below. If someone, or ourselves in certain case, want to just try to send a cancel as a best-effort thing but ignore failure I think they should catch the exception at that level.

Reply:
Looked more around and I understand that this pattern was largely copied from `_try_cancel()` which received a similar guard in b3e0be9869257cfa9602608caee4ebe96148f63f.

I think that my refactoring in that comment was somewhat misguided: the call of `_try_cancel()` from `cancel()` is already guarded by a check for connection closed (and the comment is wrong, I assume I meant signal handler, not callback handler). So I don't think I meant to make that change for the benefit of `close()`, rather I think I wanted to make the other call to `_try_cancel()` (the one in the interruption of `wait()`) a best-effort thing.

How it happens that after a Ctrl-C the connection is closed is not entirely clear to me... But I can't remember any better.

To try and recap, I don't think that `cancel()` or `cancel_safe()` should have a built-in policy of doing cancelling as as a best-effort thing. I think that these methods should just raise an exception if they fail, and in specific places we might want to ignore that exception raised and to just log a warning. So, in the code affected by b3e0be9869257cfa9602608caee4ebe96148f63f, it would be in `wait()`, but not in `cancel()`, and I haven't looked on a wider range (I haven't checked if in that moment `cancel()` was being used in other operations).

What do you think of reviewing the calls to ensure:

- `cancel()` to bail out early for closed connection, raise whatever exception
- `cancel_safe()` to do the same
- `wait()` to guard for exceptions on `close()`, log and proceed
- other uses to review.

I think that the guideline should be that if a cancelling fails when we are handling an exception, we should log the episode and continue, especially when the original exception wanted to be re-raised, rather than clobbering it with the cancelling error.

Reply:
What you suggest above is not entirely clear to me; I would agree that `cancel()` and `cancel_safe()` should behave similarly w.r.t. failures to create the underlying `PQcancel` or `PQcancelConn` object, which is why I included the exception handling here, indeed. Similarly, if something needs to be changed, it must be done consistently on both sides; but can we do this in another PR?

I can remove the `except e.OperationalError` right now though. Let me know.

Reply:
Opened #778 to look into that.